### PR TITLE
Add smart relative position

### PR DESCRIPTION
**Pull request by duanxianbo**

### CODE REVIEW

File: src/libs/layout/relativeLayout.ts [hunk 1]
```diff
@@ -4,6 +4,8 @@ export type HorizontalAlignment = 'left' | 'right' | 'center';
 
 export type VerticalAlignment = 'top' | 'bottom' | 'center';
 
+export type OverSpaceAction = 'overlap' | 'horizontalShift' | 'verticalShift';
```
Comment (naturetime10) [inline src/libs/layout/relativeLayout.ts]:
```suggestion
export type OverSpaceAction = 'overlap' | 'shiftHorizontal' | 'shiftVertical';
```

File: src/libs/layout/relativeLayout.ts [hunk 2]
```diff
@@ -50,37 +62,84 @@ function calcLeft(
...
-            return (
+            left =
                 referenceRect.left +
-                (referenceRect.width - followerRect.width) / 2
-            );
+                (referenceRect.width - followerRect.width) / 2;
+                break
+    }
+    const windowRight =
+        window.innerWidth || document.documentElement.clientWidth;
```
Comment (naturetime10) [inline src/libs/layout/relativeLayout.ts]:
```suggestion
    }
    
    const rightBound =
        (window.innerWidth || document.documentElement.clientWidth) - offset;
```

File: src/libs/layout/relativeLayout.ts [hunk 3]
```diff
@@ -50,37 +62,84 @@ function calcLeft(
...
+    const right = left + followerRect.width;
+
+    if (left >= 0 && right <= windowRight) {
```
Comment (naturetime10) [inline src/libs/layout/relativeLayout.ts]:
```suggestion
    if (left >= offset && right <= rightBound) {
```

File: src/libs/layout/relativeLayout.ts [hunk 4]
```diff
@@ -50,37 +62,84 @@ function calcLeft(
...
+
+    if (overSpaceAction !== 'horizontalShift' && left < 0) {
+        return offset
+    } else if (overSpaceAction !== 'horizontalShift' && right > windowRight) {
+        return windowRight - followerRect.width - offset
+    } else if (overSpaceAction === 'horizontalShift' && left < 0) {
+        return referenceRect.left
+    } else {
+        return referenceRect.left + referenceRect.width - followerRect.width;
     }
```
Comment (naturetime10) [inline src/libs/layout/relativeLayout.ts]: if (left < offset) {
   switch (overSpaceAction) {
      case :
         return
   }

if (right > rightBound) {
   switch (overSpaceAction) {
      case :
         return
   }
}

return left.

Let's do the same for calcTop. Hope this is making logic more clear and easier to understand.

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/layout/relativeLayout.ts (hunk 4 above)
Comment (duanxianbo) [reply src/libs/layout/relativeLayout.ts]: done

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**